subprojects/docs/src/docs/userguide/jvm/java_library_plugin.adoc

```diff
@@ -241,6 +241,146 @@ The role of each configuration is described in the following tables:
+
+Java Modules also have a version that is encoded as part of the module identity in the `module-info.class` file.
+This version can be inspect when a module is running.
+If the gradle project has a version configured, this version is used by default.
+
+.Declare the module version in the build script or directly as compile task option
+====
+include::sample[dir="snippets/java-library/module/groovy",files="build.gradle[tags=declareVersion]"]
+include::sample[dir="snippets/java-library/module/kotlin",files="build.gradle.kts[tags=declareVersion]"]
+====
+
+=== Using Libraries that are not modules
+
+You probably want to use external libraries, like OSS libraries from Maven Central, in your modular Java project.
+Some libraries, in their newer versions, are already full modules with a module descriptor.
+For example, `com.google.code.gson:gson:2.8.6` that has the module name `com.google.gson`.
+
+Others, like `org.apache.commons:commons-lang3:3.10`, may not offer a full module descriptor but will at least contain an `Automatic-Module-Name` entry in their manifest file to define the module's name (`org.apache.commons.lang3` in the example).
+Such modules, that only have a name as module description, are called _automatic module_ that export **all** their packages and can read **all** modules on the module path.
+
+A third case are traditional libraries that provide no module information at all -- for example `commons-cli:commons-cli:1.4`.
+Gradle puts such libraries on the classpath instead of the module path.
+The classpath is then treated as one module (the so called _unnamed_ module) by Java.
+
+.Dependencies to modules and libraries declared in build file
+====
+include::sample[dir="snippets/java-library/module/groovy",files="build.gradle[tags=dependencies]"]
+include::sample[dir="snippets/java-library/module/kotlin",files="build.gradle.kts[tags=dependencies]"]
+====
+
+[source,java,indent=0]
+.Module dependencies declared in module-info.java file
+----
+include::{samplesPath}/java-library/module/kotlin/src/main/java/module-info.java[]
+----
+
+While a real module cannot directly depend on the unnamed module (only by adding command line flags), automatic modules can also see the unnamed module.
+Thus, if you cannot avoid to rely on a library without module information, you can wrap that library in an automatic module as part of your project.
+How you do that is described in the next section.
+
+Another way to deal with non-modules is to enrich existing Jars with module descriptors yourself using <<artifact_transforms.adoc#,artifact transforms>>.
+link:../samples/sample_java_modules_with_transform.html[This sample] contains a small _buildSrc_ plugin registering such a transform which you may use and adjust to your needs.
+This can be interesting if you want to build a fully <<application_plugin.adoc#sec:application_modular,modular application>> and want the java runtime to treat everything as a real module, e.g. for performance considerations.
```

Comment:
Where does the performance consideration comes from at the end? Unless this can be linked to a serious resource, I would not try to guess why users would decide to be fully modular.

Reply:
Ir was an argument I often head that classloading at runtime is faster. Which makes sense if more things are more structured and theres is not big blob of classes on the classpath that see everything. But I don't feel like investing time now to research this in detail. I'll remove the note here.